ballista/scheduler/src/cluster/kv.rs

```diff
@@ -57,6 +56,8 @@ pub struct KeyValueState<
 > {
     /// Underlying `KeyValueStore`
     store: S,
+    /// ExecutorHeartbeat cache, executor_id -> ExecutorHeartbeat
+    executor_heartbeats: Arc<DashMap<String, ExecutorHeartbeat>>,
```

Comment:
For `ExecutorMetadata`, it can be cached. However, for `ExecutorData`, I don't think it should be cached, as it's frequently changed and it's hard to make it consistent with the one stored in backend storage when there are multiple active schedulers.

Actually, I don't think it's necessary to introduce multiple active schedulers in the Ballista cluster. Maybe only HA is needed. If so, we can cache the `ExecutorData` with executor available slots to avoid frequent visit to the backend storage.

Reply:
Ah, yeah I don't think `ExecutorData` is used at all anymore actually. We use `ExecutorTaskSlots` to store the task slots for storing available slots for reservation purposes. So I think we should be fine just caching `ExecutorMetadata`

> Actually, I don't think it's necessary to introduce multiple active schedulers in the Ballista cluster.

We run multiple active schedulers. The scheduler is doing a non-trivial amount of work in scheduling and it's important to us to be able to scale that layer horizontally.

Reply:
For the horizontally scaling, actually, I haven't met any bottlenecks of the scheduler after introducing the optimizations previously.

Reply:
We have :). The problem is mostly around the cost of plan serialization. We frequently see execution plans with 50k+ files and the plan is very large and causes a lot of CPU and memory overhead to serialize. 

Aside from that, the other issue is that doing zero-downtime deployments is much easier with multiple active schedulers. It is a solvable problem using leader election but for our use case it was preferable to just run multiple active schedulers and solve both the deployment and scalability issue.

Reply:
I think for the plan, it's also been cached and we only need to do the serialization only once for one query. So do you still have the problem?

The reason I do not prefer the multiple active schedulers is the contention for the execution slots, which will lead to inefficient slots updates and the future consistent hashing based task assignments.

Reply:
If the multiple active schedulers is required, I think your current design of `ExecutionReservation` is reasonable. And the consistent hashing based task assignments may be only possible for the single active scheduler without slots contention.